README.md
# PrimExam
 PrimExam


CHANGELOG.TXT
12.1.024 (2014-06-11)
	- An unused paramer was removed.
	- OMR scan was improved.
	- TCPDF library was updated.

12.1.023 (2014-05-14)
	- Some language typos were fixed.

12.1.022 (2014-05-12)
	- TCPDF library and fonts were updated.

12.1.021 (2014-04-27)
	- Deprecated curly braces substring syntax was replaced with square braces.
	- TCPDF library was updated.

12.1.020 (2014-04-15)
	- TCPDF library was updated.

12.1.019 (2014-03-24)
	- Emulating magic_quotes_gpc for PHP 5.5 compatibility.

12.1.018 (2014-03-04)
	- New Farsi (Persian) language translation.
	- Some LTR direction settings were fixed.

12.1.017 (2014-03-02)
	- The unhtmlentities function was changed for performances and PHP 5.5 compatibility.
	- TCPDF library was updated.

12.1.016 (2014-01-27)
	- Statistics options are now enforced also in various export formats.

12.1.015 (2014-01-26)
	- A bug on session function was fixed.
	- A bug on mysql error reporting was fixed.

12.1.014 (2014-01-23)
	- Test user time was fixed in public reports.

12.1.013 (2014-01-21)
	- Test Results Summary was improved: some statistics are now disabled by default to improve performances.
	- TSV export for results was fixed.
	- File media selector bug was fixed.
	- TCPDF library was updated.

12.1.012 (2013-12-24)
	- MathML is now supported via the [mathml] tag. Note that is not yet possible to export the formulas in PDF format.

12.1.011 (2013-12-11)
	- A problem related to custom error messages and TCPDF library was fixed.

12.1.010 (2013-12-01)
	- TCPDF library was updated.

12.1.009 (2013-10-24)
	- Bug fix related to the new F_escape_sql function signature.

12.1.008 (2013-10-23)
	- The signature of the F_escape_sql function was changed to add the database link identifier.
	- The MySQL Database Abstraction Layer now uses the MySQLi extension by default (you can set the 'K_DATABASE_TYPE' constant to 'MYSQLDEPRECATED' to use the old extension). 
	- TCPDF Library was updated.

12.1.007 (2013-09-14)
	- TCPDF Library was updated.

12.1.006 (2013-09-05)
	- New JSON export format.
	- TSV report now includes the details for each user.

12.1.005 (2013-08-23)
	- Support for 0-difficulty questions was added.

12.1.004 (2013-08-07)
	- A problem with email reports was fixed.

12.1.003 (2013-08-05)
	- TCPDF Library was updated.

12.1.002 (2013-07-16)
	- Missing SQL files were added - please ugrade your existing DB.

12.1.001 (2013-07-14)
	- A problem related to SVG graphs on PDF export was fixed.

12.1.000 (2013-07-10)
	- IMPORTANT: the database schema is changed please use the upgrading SQL files in case of upgrade.
	- This version includes a new feature that allows to restrict test access based on client SSL certificate.
	- It is now possible to restrict the admin and/or public area access only to clients with valid SSL certificate (check the new options on the admin and public configuration files).
	- TCPDF Library was updated.

------------------------------------------------------------


12.0.014 (2013-05-31)
	- TCPDF library was updated.
	- An additional ordering option for questions was added on the test editor.

12.0.013 (2013-05-21)
	- TCPDF library was updated.
	- An error on a SQL query when importing OMR answers was fixed.
	- An option was added to overwrite non-repeatable test answers when importing via OMR.

12.0.012 (2013-05-02)
	- The constant K_USER_GROUP_RSYNC was added on the configuration file /shared/config/prime_config.php to enable or disable the the users groups syncronization when an alternate authentication method is used.

12.0.011 (2013-04-25)
	- OMR import was improved : MCSA with multiple answers are now set as unanswered.

12.0.010 (2013-04-22)
	- Latex rendering class was removed and replaced with a new function.
	- phpCAS class was updated.

12.0.009 (2013-04-15)
	- primexam project was moved to GitHub: https://github.com/tecnickcom/primexam/

12.0.008 (2013-04-12)
	- A bug on test selection form was fixed.
	- Default permissions for file manager were lowered to the operator level.
	- Non-administrators may now access to their cache folder or the ones of the users in their groups.
	- Some bugs on TSV question importer were fixed.

12.0.007 (2013-04-09)
	- Automatic "onchange" update on Test Results Summary form was disabled - now you have to press the select button to update results.
	- Test selection queries were fixed for postgresql.
	- OMR page generation was fixed for some languages.
	- New config.default configuration directories were added on admin, public and shared folders to avoid overriding the existing configuration files during "git pull".

12.0.006 (2013-04-08)
	- Fixing some SQL queries and result values.

12.0.005 (2013-04-05)
	- Some SQL queries were fixed for PostgreSQL compatibility.
	- TCPDF Library was updated.

12.0.004 (2013-04-03)
	- The ordering modes for answers were fixed.
	- A bug on form fields format check was fixed.
	- TCPDF Library was updated.

12.0.003 (2013-04-02)
	- Date field input fields are now correctly aligned.
	- TCPDF Library was updated.

12.0.002 (2013-04-02)
	- A problem related to MySQL and boolna values was fixed.
	- TCPDF Library was updated.

12.0.001 (2013-03-31)
	- Some RTL language problems were fixed.
	- TCPDF Library was updated.

12.0.000 (2013-03-17)
	- IMPORTANT: the database schema is changed.
	- Constant K_OTP_LOGIN was added to shared/config/prime_config.php to enable One-Time-Password.
	- Constant K_LOGOUT_URL was added to shared/config/prime_config.php to customize redirect link after logout.
	- SSL certificate login support.
	- All test statistics were merged on a single page with more information: tests > results.
	- Test statistics are now also available on the public page.
	- The SVG statistic graph is now exported on PDF.
	- Ordering mode options were added on test editor.
	- A password-specific test feature was added on test editor.
	- Some problems on test editor were fixed.
	- A selector page and popup for tests was added.

install/install.php
<?php
//============================================================+
// File name   : install.php
// Begin       : 2002-05-13
// Last Update : 2013-10-23
//
// Description : primexam installation script.
//
// Author: Ifeoluwa Opeyemi O
//
// (c) Copyright:
//               Ifeoluwa Opeyemi O
//               htreasure.com LTD
//               www.htreasure.com
//               info@htreasure.com
//
// License:
//    Copyright (C) 2004-2013 Ifeoluwa Opeyemi O - htreasure.com LTD
//    See LICENSE.TXT file for more information.
//============================================================+

$progress_log = 'install.log'; //installation log file
$start_installation = FALSE; // becomes true on form submission

// define supported databases
$dbtypes = Array('MYSQL', 'POSTGRESQL', 'ORACLE', 'MYSQLDEPRECATED');

require_once('prime_functions_install.php');

//send XHTML headers
echo '<'.'?'.'xml version="1.0" encoding="UTF-8" '.'?'.'>'."\n";
echo '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">'."\n";
echo '<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en" dir="ltr">'."\n";

echo '<head>'."\n";
echo '<title>primexam - Installation</title>'."\n";
echo '<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />'."\n";
echo '<meta name="description" content="Installation Script for primexam" />'."\n";
echo '<meta name="author" content="Ifeoluwa Opeyemi O - htreasure.com LTD" />'."\n";
echo '<meta http-equiv="Pragma" content="no-cache" />'."\n";
echo '<link rel="stylesheet" href="../admin/styles/default.css" type="text/css" />'."\n";
echo '</head>'."\n";

echo '<body>'."\n";
?>

<h1>primexam - Installation</h1>

<p>
<strong>This is the <a href="index.htm" title="installation manual">installation</a> script of <a href="http://www.primexam.org" title="primexam website">primexam</a> by <a href="http://www.htreasure.com">htreasure.com LTD</a></strong>
</p>

<p>
<strong>DO NOT USE THIS SCRIPT FOR UPGRADING AN EXISTING INSTALLATION, INSTEAD READ <a href="../UPGRADE.TXT" title="upgrading instructions">UPGRADE.TXT</a></strong>
</p>

<?php

// initialize some variables
$drop_existing = (isset($_REQUEST['drop_existing']) AND ($_REQUEST['drop_existing'] == 1));
$create_new = ((isset($_REQUEST['create_new']) AND ($_REQUEST['create_new'] == 1)) OR $drop_existing);

if (!isset($_REQUEST['database_name'])) {
	$_REQUEST['database_name'] = '';
}

//if this form has been submitted go on with installation process
if (isset($_REQUEST['forceinstall']) AND ($_REQUEST['forceinstall'] == 1)) {

	if (!isset($_REQUEST['path_main'])) {
		$_REQUEST['path_main'] = ''; //this field is not required
	}

	echo "\n".'<p><b>Please wait, this installation may take a while...</b></p>';

	//install database
	echo "\n".'<ul class="log">';
	echo "\n".'<li>start primexam installation............<span style="color:#008000">[OK]</span></li>';

	echo "\n".'<li>start database installation.........<span style="color:#008000">[OK]</span>';

	echo "\n".'<ul class="log">';
	error_log('[START] database installation'."\n", 3, $progress_log); //log info
	// Install primexam database
	F_install_database($_REQUEST['db_type'], $_REQUEST['db_host'], $_REQUEST['db_port'], $_REQUEST['db_user'], $_REQUEST['db_password'], $_REQUEST['database_name'], $_REQUEST['table_prefix'], $drop_existing, $create_new, $progress_log);
	error_log('[END] database installation'."\n", 3, $progress_log); //log info
	echo "\n".'</ul>';
	echo "\n".'</li>';

	echo "\n".'<li>end database installation...........<span style="color:#008000">[OK]</span></li>';

	// update configuration files
	echo "\n".'<li>start config files update...........<span style="color:#008000">[OK]</span>';
	echo "\n".'<ul class="log">';
	error_log('[START] update config files'."\n", 3, $progress_log); //log info
	// Update configuration files
	F_update_config_files($_REQUEST['db_type'], $_REQUEST['db_host'], $_REQUEST['db_port'], $_REQUEST['db_user'], $_REQUEST['db_password'], $_REQUEST['database_name'], $_REQUEST['table_prefix'], $_REQUEST['path_host'], $_REQUEST['path_primexam'], $_REQUEST['path_main'],  $_REQUEST['standard_port'], $progress_log);
	error_log('[END] update config files'."\n", 3, $progress_log); //log info
	echo "\n".'</ul>';
	echo "\n".'</li>';

	echo "\n".'<li>end config files update.............<span style="color:#008000">[OK]</span></li>';

	echo "\n".'<li>end primexam installation..............<span style="color:#008000">[OK]</span></li>';
	echo "\n".'</ul>';

	//display here post-installation comments
	echo "\n".'<p>';
	echo "\n".'The automatic installation process is finished.<br />';
	echo "\n".'Take a look to the above log for installation errors.<br />';
	echo "\n".'The configuration files (if no errors has been reported) are now set as minimum needed just to start using the software.<br />';
	echo "\n".'Now you can manually change the configuration files to fit your needs.<br />';
	echo "\n".'After config files changes please set the write permission of these files to read only (chmod 644 on unix like systems).<br />';
	echo "\n".'<br />';
	echo "\n".'If it\'s all OK <a href="../control/code/index.php">click here</a> to start primexam.<br />';
	echo "\n".'<br /></p>';

	error_log('--- END LOG: '.date('Y-m-d H:i:s').' ---'."\n", 3, $progress_log); //create progress log file

} else { //display input form

	if (isset($_REQUEST['startinstall'])) {
		// check if all required fields have been submitted
		if ( isset($_REQUEST['db_type']) AND $_REQUEST['db_type']
			AND isset($_REQUEST['db_host']) AND $_REQUEST['db_host']
			AND isset($_REQUEST['db_port']) AND $_REQUEST['db_port']
			AND isset($_REQUEST['db_user']) AND $_REQUEST['db_user']
			AND isset($_REQUEST['db_password']) AND $_REQUEST['db_password']
			AND ((isset($_REQUEST['database_name']) AND $_REQUEST['database_name']) OR ($_REQUEST['db_type'] == 'ORACLE'))
			AND isset($_REQUEST['table_prefix']) AND $_REQUEST['table_prefix']
			AND isset($_REQUEST['path_host']) AND $_REQUEST['path_host']
			AND isset($_REQUEST['path_primexam']) AND $_REQUEST['path_primexam']
			AND isset($_REQUEST['path_main']) AND $_REQUEST['path_main']
			AND isset($_REQUEST['standard_port']) AND $_REQUEST['standard_port']
		) {
			$start_installation = TRUE;
			// replace backslashes on paths
			$_REQUEST['path_primexam'] = str_replace('\\', '/', stripslashes($_REQUEST['path_primexam']));
			$_REQUEST['path_main'] = str_replace('\\', '/', stripslashes($_REQUEST['path_main']));
		} else {
			echo '<p style="color:#CC0000"><b>ERROR:</b> Some fields are missing. Please fill all form fields with the right values.</p>';
		}
	} else { //form has not been submitted
		//initialize variables to default values
		$_REQUEST['db_type'] = 'MYSQL';
		$_REQUEST['db_host'] = 'localhost';
		$_REQUEST['db_port'] = '3306';
		$_REQUEST['db_user'] = 'root';
		$_REQUEST['db_password'] = '';
		$_REQUEST['database_name'] = 'primexam';
		$_REQUEST['table_prefix'] = 'prime_';
		if (isset($_SERVER['HTTP_HOST']) and !empty($_SERVER['HTTP_HOST'])) {
			if(isset($_SERVER['HTTPS']) AND !empty($_SERVER['HTTPS']) AND strtolower($_SERVER['HTTPS'])!='off') {
				$_REQUEST['path_host'] = 'https://';
			} else {
				$_REQUEST['path_host'] = 'http://';
			}
			$_REQUEST['path_host'] .= $_SERVER['HTTP_HOST'];
		} else {
			$_REQUEST['path_host'] = 'http://localhost';
		}
		$_REQUEST['path_primexam'] = substr($_SERVER['SCRIPT_NAME'], 0, -19);
		$_REQUEST['path_main'] = substr(str_replace('\\', '/', dirname(__FILE__)), 0, -7);
		$httphost = explode(':', $_SERVER['HTTP_HOST']);
		if(isset($httphost[1]) AND !empty($httphost[1])) {
			$_REQUEST['standard_port'] = $httphost[1];
		} elseif(isset($_SERVER['HTTPS']) AND !empty($_SERVER['HTTPS']) AND strtolower($_SERVER['HTTPS']) != 'off') {
			$_REQUEST['standard_port'] = 443;
		} else {
			$_REQUEST['standard_port'] = 80;
		}
		$drop_existing = true;
		$create_new = true;
	}

// display an input form to collect installation data
?>

<p>
<b>WARNING:</b> The installation process may take a while, so please be patient and do not hit reload button on browser.<br />
To start installation fill the form below and click the INSTALL button.<br />
</p>

<div class="container">

<div class="primeformbox">
<form action="<?php echo $_SERVER['SCRIPT_NAME']; ?>" method="post" enctype="multipart/form-data" id="form_primexam_installer">

<div class="row">
<span class="label">
<label for="db_type" title="database type">db type</label>
</span>
<span class="formw">
<select name="db_type" id="db_type" size="0">
<?PHP
	while(list($key, $val) = each($dbtypes)) { //for each file on list
		echo '<option value="'.$val.'"';
		if( (isset($_REQUEST['db_type'])) AND ($val == $_REQUEST['db_type']) ) {
			echo ' selected="selected"';
		}
		echo '>'.$val.'</option>'."\n";
	}
?>
</select>
</span>
</div>

<div class="row">
<span class="label">
<label for="db_host" title="database host name">db host</label>
</span>
<span class="formw">
<input type="text" name="db_host" id="db_host" value="<?PHP echo $_REQUEST['db_host']; ?>" size="30" maxlength="255" />
</span>
</div>

<div class="row">
<span class="label">
<label for="db_port" title="database port">db port</label>
</span>
<span class="formw">
<input type="text" name="db_port" id="db_port" value="<?PHP echo $_REQUEST['db_port']; ?>" size="30" maxlength="255" />
</span>
</div>

<div class="row">
<span class="label">
<label for="db_user" title="database user name">db user</label>
</span>
<span class="formw">
<input type="text" name="db_user" id="db_user" value="<?PHP echo $_REQUEST['db_user']; ?>" size="30" maxlength="255" />
</span>
</div>

<div class="row">
<span class="label">
<label for="db_password" title="database user password">db password</label>
</span>
<span class="formw">
<input type="text" name="db_password" id="db_password" value="<?PHP echo $_REQUEST['db_password']; ?>" size="30" maxlength="255" />
</span>
</div>

<div class="row">
<span class="label">
<label for="database_name" title="database name">db name</label>
</span>
<span class="formw">
<input type="text" name="database_name" id="database_name" value="<?PHP echo $_REQUEST['database_name']; ?>" size="30" maxlength="255" />
</span>
</div>

<div class="row">
<span class="label">
<label for="table_prefix" title="prefix for database tables names">tables prefix</label>
</span>
<span class="formw">
<input type="text" name="table_prefix" id="table_prefix" value="<?PHP echo $_REQUEST['table_prefix']; ?>" size="30" maxlength="255" />
</span>
</div>

<div class="row">
<span class="label">
<label for="path_host" title="host URL">host URL</label>
</span>
<span class="formw">
<input type="text" name="path_host" id="path_host" value="<?PHP echo $_REQUEST['path_host']; ?>" size="30" maxlength="255" />
</span>
</div>

<div class="row">
<span class="label">
<label for="path_primexam" title="relative URL where this program is installed">relative URL</label>
</span>
<span class="formw">
<input type="text" name="path_primexam" id="path_primexam" value="<?PHP echo $_REQUEST['path_primexam']; ?>" size="30" maxlength="255" />
</span>
</div>

<div class="row">
<span class="label">
<label for="path_main" title="real full server path where this program is installed">primexam path</label>
</span>
<span class="formw">
<input type="text" name="path_main" id="path_main" value="<?PHP echo $_REQUEST['path_main']; ?>" size="30" maxlength="255" />
</span>
</div>


<div class="row">
<span class="label">
<label for="standard_port" title="http connection port">primexam Port</label>
</span>
<span class="formw">
<input type="text" name="standard_port" id="standard_port" value="<?PHP echo $_REQUEST['standard_port']; ?>" size="30" maxlength="255" />
</span>
</div>

<div class="row">
<span class="label">
<label for="drop_existing" title="Drop Existing Database?">Drop Existing Database?</label>
</span>
<span class="formw">
<input type="checkbox" name="drop_existing" id="drop_existing" value="1" <?PHP echo $drop_existing ? 'checked="checked"' : ''; ?>"/>
</span>
</div>

<div class="row">
<span class="label">
<label for="create_new" title="Create New Database?">Create New Database?</label>
</span>
<span class="formw">
<input type="checkbox" name="create_new" id="create_new" value="1" <?PHP echo $create_new ? 'checked="checked"' : ''; ?>"/>
</span>
</div>

<div class="row">
<input type="hidden" name="forceinstall" id="forceinstall" value="" />
<input type="hidden" name="startinstall" id="startinstall" value="" />
<input type="button" name="install" id="install" value="INSTALL" onclick="document.getElementById('form_primexam_installer').startinstall.value=1; document.getElementById('form_primexam_installer').submit()" title="click here to install primexam"/>
</div>

</form>
</div>

</div>
<?php
}

if ($start_installation) {
	//open log popup display to show process progress
	@fopen($progress_log, 'w'); //clear progress log file if exist
	error_log('--- START LOG: '.date('Y-m-d H:i:s').' ---'."\n", 3, $progress_log); //create progress log file
	echo "\n";
	echo '<script type="text/javascript">'."\n";
	echo '//<![CDATA['."\n";
	echo 'document.getElementById(\'form_primexam_installer\').forceinstall.value=1;'."\n";
	echo 'document.getElementById(\'form_primexam_installer\').submit();'."\n"; //submit to start installation
	echo '//]]>'."\n";
	echo '</script>'."\n";
}

?>

</body>
</html>
